video/out/wayland_common.c

```diff
@@ -2898,19 +3504,39 @@ static void set_content_type(struct vo_wayland_state *wl)
 static void set_cursor_shape(struct vo_wayland_seat *s)
 {
 #if HAVE_WAYLAND_PROTOCOLS_1_32
-    wp_cursor_shape_device_v1_set_shape(s->cursor_shape_device, s->pointer_enter_serial,
-                                        WP_CURSOR_SHAPE_DEVICE_V1_SHAPE_DEFAULT);
+    if (s->cursor_shape_device)
+        wp_cursor_shape_device_v1_set_shape(s->cursor_shape_device, s->pointer_enter_serial,
+                                            WP_CURSOR_SHAPE_DEVICE_V1_SHAPE_DEFAULT);
+
+    struct vo_wayland_tablet_tool *tablet_tool;
+    wl_list_for_each(tablet_tool, &s->tablet_tool_list, link) {
+        if (tablet_tool->cursor_shape_device)
+            wp_cursor_shape_device_v1_set_shape(tablet_tool->cursor_shape_device,
+                                                tablet_tool->proximity_serial,
+                                                WP_CURSOR_SHAPE_DEVICE_V1_SHAPE_DEFAULT);
+    }
 #endif
 }
 
+static void set_cursor(struct vo_wayland_seat *s, struct wl_surface *cursor_surface, int32_t hotspot_x, int32_t hotspot_y) {
+    wl_pointer_set_cursor(s->pointer, s->pointer_enter_serial,
+                          cursor_surface, hotspot_x, hotspot_y);
+
+    struct vo_wayland_tablet_tool *tablet_tool;
+    wl_list_for_each(tablet_tool, &s->tablet_tool_list, link) {
+        zwp_tablet_tool_v2_set_cursor(tablet_tool->tablet_tool, tablet_tool->proximity_serial,
+                                      cursor_surface, hotspot_x, hotspot_y);
+    }
+}
+
 static int set_cursor_visibility(struct vo_wayland_seat *s, bool on)
 {
     if (!s)
         return VO_FALSE;
     struct vo_wayland_state *wl = s->wl;
     wl->cursor_visible = on;
     if (on) {
-        if (s->cursor_shape_device) {
+        if (s->wl->cursor_shape_manager) {
```

Comment:
The somewhat tricky part: is it safe to assume that having `s->wl->cursor_shape_manager` at this point results in a valid `s->cursor_shape_device` at this point? I'm still checking that within `set_cursor_shape()` just in case that there is no pointer but only a tablet, but we also don't hit the fallback path in that case.

Reply:
I believe so. The `cursor_shape_device` gets created as part of the `seat_handle_caps` events. During the initial vo init, we specifically run a couple of `wl_display_roundtrip` calls to ensure all of the wayland objects get initialized before anything else happens. This is before `set_cursor_visibility_all_seats` can get called by mpv's playback loop. The only two wrenches here are the pointer enter and and tablet proximity in events also calling this function. I can't see how any compositor would give us that before the seat events though. That doesn't sound like it could work so it is probably OK.